File: Greedy/Java/Prime Subtraction Operation.java
// Problem link: https://leetcode.com/problems/prime-subtraction-operation/description
// Time: O(n * sqrt(m))
// Space: O(max(nums))
class Solution {
    public boolean primeSubOperation(int[] nums) {
        int max = nums[0], prev = 0, upper, largestPrime;

        for(int num : nums) {
            max = Math.max(num, max);
        }

        int[] primes = new int[max + 1];

        for(int num = 2; num <= max; num++) {
            if(this.isPrime(num)) {
                primes[num] = num;
            }
            else {
                primes[num] = primes[num - 1];
            }
        }

        for(int num : nums) {
            upper = num - prev;
            largestPrime = upper > 1 ? primes[upper - 1] : 0;

            if(num - largestPrime <= prev) {
                return false;
            }

            prev = num - largestPrime;
        }

        return true;
    }

    private boolean isPrime(int num) {
        for(int n = 2; n <= Math.sqrt(num); n++) {
            if(num % n == 0) {
                return false;
            }
        }

        return true;
    }
}

File: Greedy/Python/Prime Subtraction Operation.py
# Problem link: https://leetcode.com/problems/prime-subtraction-operation/description
# Time: O(n * sqrt(m))
# Space: O(max(nums))
class Solution:
    def primeSubOperation(self, nums: List[int]) -> bool:
        primes, prev = [0, 0], 0

        for num in range(2, max(nums) + 1):
            if self.isPrime(num):
                primes.append(num)
            else:
                primes.append(primes[-1])
        
        for num in nums:
            upper = num - prev
            largestPrime = primes[upper - 1]

            if num - largestPrime <= prev:
                return False
            prev = num - largestPrime

        return True
    
    def isPrime(self, num):
        for n in range(2, int(sqrt(num)) + 1):
            if num % n == 0:
                return False
        return True
